src/Symfony/Component/HttpClient/RetryHttpClient.php

```diff
@@ -0,0 +1,114 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\Response\AsyncContext;
+use Symfony\Component\HttpClient\Response\AsyncResponse;
+use Symfony\Component\HttpClient\Response\MockResponse;
+use Symfony\Component\HttpClient\Retry\ExponentialBackOff;
+use Symfony\Component\HttpClient\Retry\HttpCodeDecider;
+use Symfony\Component\HttpClient\Retry\RetryBackOffInterface;
+use Symfony\Component\HttpClient\Retry\RetryDeciderInterface;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Automatically retries the failing requests.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RetryHttpClient implements HttpClientInterface
+{
+    use AsyncDecoratorTrait;
+
+    private $decider;
+    private $strategy;
+    private $maxRetries;
+    private $logger;
+
+    /**
+     * @param int $maxRetries The maximum number of times to retry
+     */
+    public function __construct(HttpClientInterface $client, RetryDeciderInterface $decider = null, RetryBackOffInterface $strategy = null, int $maxRetries = 3, LoggerInterface $logger = null)
+    {
+        $this->client = $client;
+        $this->decider = $decider ?? new HttpCodeDecider();
+        $this->strategy = $strategy ?? new ExponentialBackOff();
+        $this->maxRetries = $maxRetries;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        $retryCount = 0;
+
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount) {
+            $exception = null;
+            try {
+                if ($chunk->isTimeout() || null !== $chunk->getInformationalStatus()) {
+                    yield $chunk;
+
+                    return;
+                }
+
+                // only retry first chunk
+                if (!$chunk->isFirst()) {
+                    $context->passthru();
+                    yield $chunk;
+
+                    return;
+                }
+            } catch (TransportExceptionInterface $exception) {
+                // catch TransportExceptionInterface to send it to strategy.
+            }
+
+            if ($retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $partialResponse = new MockResponse($context->getContent(), ['http_code' => $statusCode = $context->getStatusCode(), 'headers' => $headers = $context->getHeaders()]), $exception)) {
```

Comment:
style-wise, $statusCode+$headers variables could be assigned before the if

Reply:
when `retryCount >= $this->maxRetries` those variable are not needed. It's not really an issue for $status, but it avoid fetching/parsing headers for nothing.

I can leverage `$partialResponse` instead of $status and $headers if you prefer

Reply:
I'm not sure it's worth optimizing this: retries are by definition rare.